logger/plotter.py

```diff
@@ -0,0 +1,102 @@
+import numpy as np
+import pprint
+
+from collections import defaultdict
+
+# optional visdom
+try:
+    import visdom
+except ImportError:
+    visdom = None
+
+
+class Cache(object):
+    def __init__(self):
+        self.clear()
+
+    def clear(self):
+        self._x = []
+        self._y = []
+
+    def update(self, metric):
+        self._x.append(metric.index.get())
+        self._y.append(metric.get())
+
+    @property
+    def x(self):
+        return np.array(self._x)
+
+    @property
+    def y(self):
+        return np.array(self._y)
+
+
+class Plotter(object):
+
+    def __init__(self, xp, visdom_opts):
+        super(Plotter, self).__init__()
+
+        if visdom_opts is None:
+            visdom_opts = {}
+
+        assert visdom is not None, "visdom could not be imported"
+
+        # visdom env is given by Experiment name unless specified
+        if 'env' not in list(visdom_opts.keys()):
+            visdom_opts['env'] = xp.name
+
+        self.viz = visdom.Visdom(**visdom_opts)
+        self.windows = {}
+        self.append = {}
+        self.cache = defaultdict(Cache)
+
+    def _plot_xy(self, name, tag, x, y, time_idx=True):
+        """
+        Creates a window if it does not exist yet.
+        Returns True / non-empty string if data has been sent successfully,
+        False otherwise.
+        """
+        xlabel = 'Time (s)' if time_idx else 'Index'
+        if name not in list(self.windows.keys()):
+            opts = {'legend': [tag], 'title': name, 'xlabel': xlabel}
+            self.windows[name] = self.viz.line(Y=y, X=x, opts=opts)
+            return True
+        else:
+            return self.viz.updateTrace(Y=y, X=x, name=tag,
```

Comment:
So this returns False if it fails? Not hang?

Reply:
Actually yes, the error is caught [there](https://github.com/facebookresearch/visdom/blob/master/py/__init__.py#L266)